trulens_eval/examples/snowflake_demo/schema.py

```diff
@@ -41,23 +51,73 @@ def __init__(self):
         self.model_config: ModelConfig = ModelConfig()
         self.feedback: ConversationFeedback = None
 
-    def add_message(self, message: Message, container=None, render=True):
+    def add_message(self, message: Message, container=st, render=True):
         self.messages.append(message)
         if render:
-            self.render_message(message, container)
+            self.render_message(message, container, key=str(len(self.messages)))
 
     def reset_messages(self):
         self.messages: List[Message] = []
 
-    def render_all(self, container=None):
-        for message in self.messages:
-            self.render_message(message, container)
+    def render_all(self, container=st):
+        for idx, message in enumerate(self.messages):
+            self.render_message(message, container, key=str(idx))
+
+    @st.experimental_fragment(run_every=2)
```

Comment:
this is super cool, had just learned about this streamlit feature last week at summit

Reply:
haha yeah, the feature definitely has some of its rough edges though - took me too long to figure out that a key is needed to aid streamlit in rendering the screen